{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1b638131-ee12-45bd-9687-9dcfac5afd49",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The new embeddings will be initialized from a multivariate normal distribution that has old embeddings' mean and covariance. As described in this article: https://nlp.stanford.edu/~johnhew/vocab-expansion.html. To disable this, use `mean_resizing=False`\n",
      "/tmp/ipykernel_114/992382303.py:15: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.\n",
      "  state = torch.load(\"genechat_checkpoints/model_best.pt\", map_location=device)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=== Example 1 Ground Truth ===\n",
      "This genomic region was validated as an active enhancer by the ChIP-STARR-seq massively parallel reporter assay in naive human embryonic stem cells. This enhancer is marked by the H3K27ac and H3K4me1 histone modifications.\n",
      "\n",
      "=== Example 1 Model Output ===\n",
      "of insulin secretion; and negative regulation of signal transduction. Predicted to be located in cytoplasm. Predicted to be active in plasma membrane. Is expressed in intestinal bulb. Orthologous to human SLC17A1 (solute carrier family 17 member 1). Human ortholog(\n",
      "\n",
      "=== Example 2 Ground Truth ===\n",
      "This genomic sequence was predicted to be a transcriptional regulatory region based on chromatin state analysis from the ENCODE (ENCyclopedia Of DNA Elements) project. It was validated as a functional repressive element by the Sharpr-MPRA technique (Systematic high-resolution activation and repression profiling with reporter tiling using massively parallel reporter assays) in K562 erythroleukemia cells (group: K562 Repressive DNase unmatched - State 12:CtcfO, distal CTCF/candidate insulator with open chromatin).\n",
      "\n",
      "=== Example 2 Model Output ===\n",
      "transcription. Acts upstream of or within negative regulation of cell population proliferation. Located in nucleus. Is expressed in alimentary system; heart; integument; nervous system; and respiratory system. Orthologous to several human genes including PRR26 (PRR family 26 subfamily B member). Orthologous\n"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "from dataset import read_data\n",
    "from genechat_model import GeneChatModel\n",
    "\n",
    "device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
    "\n",
    "train_data, test_data, _, _ = read_data()\n",
    "\n",
    "model = GeneChatModel(\n",
    "    gene_chunk_nt=512,\n",
    "    gene_chunk_overlap=0,\n",
    "    freeze_gene_encoder=True,\n",
    ").to(device)\n",
    "\n",
    "state = torch.load(\"genechat_checkpoints/model_best.pt\", map_location=device)\n",
    "model.load_state_dict(state)\n",
    "\n",
    "example1 = test_data[10]\n",
    "example2 = test_data[200]\n",
    "\n",
    "for ex_i, ex in enumerate([example1, example2], start=1):\n",
    "    dna = ex[\"dna\"]\n",
    "    target = ex[\"target\"]\n",
    "    print(f\"\\n=== Example {ex_i} Ground Truth ===\")\n",
    "    print(target)\n",
    "\n",
    "    gen = model.generate(\n",
    "        dna=dna,\n",
    "        max_new_tokens=80,\n",
    "        device=device,\n",
    "        temperature=0.8,\n",
    "        top_k=50,\n",
    "    )\n",
    "    print(f\"\\n=== Example {ex_i} Model Output ===\")\n",
    "    print(gen)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5dd5b3f9-4459-4649-b543-4afa400f138a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# model.eval()\n",
    "# !pip install sacrebleu rouge-score nltk matplotlib tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f9459545-b4f6-4458-9788-7ae85084383f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[nltk_data] Downloading package wordnet to /home/jovyan/nltk_data...\n",
      "[nltk_data]   Package wordnet is already up-to-date!\n"
     ]
    }
   ],
   "source": [
    "import nltk\n",
    "nltk.download('wordnet', download_dir='/home/jovyan/nltk_data')\n",
    "nltk.data.path.append('/home/jovyan/nltk_data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "3718a01a-caad-4959-80e1-09d4ca5be028",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- Sample idx 1242 ---\n",
      "Ground Truth: This gene encodes a homeobox-containing protein that belongs to the NK-2 homeobox family. This protein is a vertebrate homolog of Drosophila homeobox-containing protein called 'tinman', which has been...\n",
      "Prediction:   be involved in protein methylation. Predicted to be located in cytoplasm and nucleus. Human ortholog(s) of this gene implicated in methylmalonic aciduria. Orthologous to human HMTM1 (methylmalonyl-CoA...\n",
      "Metrics: {'bleu': 1.3772734577189594, 'meteor': 0.11492947386638977, 'rougeL': 0.13333333333333333}\n",
      "\n",
      "--- Sample idx 197 ---\n",
      "Ground Truth: This gene encodes a member of the DEAD box protein family. These proteins are characterized by the conserved motif Asp-Glu-Ala-Asp (DEAD) and are putative RNA helicases. They are implicated in a numbe...\n",
      "Prediction:   in nucleus. Orthologous to human NFKBIE (NFKB inhibitor epsilon). Orthologous to human NFKBIE1 (NFKB inhibitor epsilon 1). Orthologous to human KNMT1 (NFKB inhibitor epsilon 1). Orthologous to human...\n",
      "Metrics: {'bleu': 0.09043356012319995, 'meteor': 0.012335526315789474, 'rougeL': 0.03658536585365854}\n",
      "\n",
      "--- Sample idx 1750 ---\n",
      "Ground Truth: Killer cell immunoglobulin-like receptors (KIRs) are transmembrane glycoproteins expressed by natural killer cells and subsets of T cells. The KIR genes are polymorphic and highly homologous and they ...\n",
      "Prediction:   activity. Acts upstream of or within several processes, including negative regulation of neuron apoptotic process; regulation of neuron population proliferation; and regulation of neuron apoptotic pro...\n",
      "Metrics: {'bleu': 0.11223502220036934, 'meteor': 0.060960853578368696, 'rougeL': 0.07758620689655171}\n",
      "\n",
      "--- Sample idx 1316 ---\n",
      "Ground Truth: KRT12 encodes the type I intermediate filament chain keratin 12, expressed in corneal epithelia. Mutations in this gene lead to Meesmann corneal dystrophy....\n",
      "Prediction:   to human ZNF771 (zinc finger protein 771). Orthologous to human ZNF771L (zinc finger protein 771 like). Orthologous to human ZNF771 (zinc finger protein 771 like). Orthologous to human ZNF771L2 (zinc...\n",
      "Metrics: {'bleu': 1.2090878038257866, 'meteor': 0.02109704641350211, 'rougeL': 0.03773584905660378}\n",
      "\n",
      "--- Sample idx 1527 ---\n",
      "Ground Truth: This gene belongs to the peroxin-12 family. Peroxins (PEXs) are proteins that are essential for the assembly of functional peroxisomes. The peroxisome biogenesis disorders (PBDs) are a group of geneti...\n",
      "Prediction:   ; positive regulation of autophagy; and positive regulation of vesicle-mediated transport. Located in membrane. Is expressed in several structures, including alimentary system; brain; genitourinary sy...\n",
      "Metrics: {'bleu': 0.22437740376979684, 'meteor': 0.03326996197718631, 'rougeL': 0.06711409395973154}\n",
      "\n",
      "--- Sample idx 1244 ---\n",
      "Ground Truth: This gene encodes a DEAD box protein. DEAD box proteins, characterized by the conserved motif Asp-Glu-Ala-Asp (DEAD), are putative RNA helicases. They are implicated in a number of cellular processes ...\n",
      "Prediction:   is involved in the formation of the cytoskeleton of the cell and in the establishment of mitotic nuclear bodies. Mutations in this gene are associated with autosomal recessive nonsyndromic deafness. A...\n",
      "Metrics: {'bleu': 1.1461418993777734, 'meteor': 0.13595669255780346, 'rougeL': 0.11764705882352941}\n",
      "\n",
      "--- Sample idx 3947 ---\n",
      "Ground Truth: Predicted to enable K63-linked deubiquitinase activity and cysteine-type deubiquitinase activity. Predicted to be involved in negative regulation of canonical Wnt signaling pathway; positive regulatio...\n",
      "Prediction:   RNA polymerase II cis-regulatory region sequence-specific DNA binding activity. Predicted to be involved in regulation of transcription by RNA polymerase II. Predicted to be active in nucleus. Is expr...\n",
      "Metrics: {'bleu': 13.40765885824344, 'meteor': 0.14833193509808473, 'rougeL': 0.21978021978021978}\n",
      "\n",
      "--- Sample idx 1686 ---\n",
      "Ground Truth: The protein encoded by this gene is a membrane mannose-specific lectin that cycles between the endoplasmic reticulum, endoplasmic reticulum-Golgi intermediate compartment, and cis-Golgi, functioning a...\n",
      "Prediction:   The encoded preproprotein is proteolytically processed to generate a mature proteoglycansoglycan that is cleaved into two mature proteoglycansoglycans. This proteoglycan is activated by phagocytic enz...\n",
      "Metrics: {'bleu': 0.8282930259489594, 'meteor': 0.09001007332950618, 'rougeL': 0.14754098360655737}\n",
      "\n",
      "--- Sample idx 2889 ---\n",
      "Ground Truth: Predicted to enable metal ion binding activity. Predicted to be involved in peptidyl-diphthamide biosynthetic process from peptidyl-histidine and tRNA wobble uridine modification. Predicted to be loca...\n",
      "Prediction:   Predicted to be active in nucleus. Orthologous to human TSSD1 (testis specific serine/threonine kinase domain containing 1). Orthologous to human TSSD2 (testis specific serine/threonine kinase domain ...\n",
      "Metrics: {'bleu': 9.222694259817908, 'meteor': 0.13267223382045928, 'rougeL': 0.21686746987951808}\n",
      "\n",
      "--- Sample idx 3943 ---\n",
      "Ground Truth: Predicted to enable protein serine/threonine kinase activity. Predicted to be involved in negative regulation of apoptotic process and regulation of mitotic cell cycle. Predicted to be active in cytop...\n",
      "Prediction:   in protein O-linked glycosylation. Is active in cytosol and plasma membrane. Orthologous to human TGM2 (transglutaminase 2). Is expressed in brain; digestive system; immature eye; musculature system; ...\n",
      "Metrics: {'bleu': 2.548807901150031, 'meteor': 0.15306122448979592, 'rougeL': 0.18750000000000003}\n"
     ]
    }
   ],
   "source": [
    "NUM_SAMPLES = 10\n",
    "random_indices = random.sample(range(len(test_data)), NUM_SAMPLES)\n",
    "metrics_list = []\n",
    "\n",
    "for idx in random_indices:\n",
    "    ex = test_data[idx]\n",
    "    dna = ex[\"dna\"]\n",
    "    ref = ex[\"target\"]\n",
    "\n",
    "    pred = model.generate(\n",
    "        dna=dna,\n",
    "        max_new_tokens=80,\n",
    "        device=device,\n",
    "        temperature=0.7,\n",
    "        top_k=50,\n",
    "    )\n",
    "\n",
    "    m = compute_metrics(ref, pred)\n",
    "    metrics_list.append(m)\n",
    "\n",
    "    print(f\"\\n--- Sample idx {idx} ---\")\n",
    "    print(\"Ground Truth:\", ref[:200].replace(\"\\n\",\" \") + \"...\")\n",
    "    print(\"Prediction:  \", pred[:200].replace(\"\\n\",\" \") + \"...\")\n",
    "    print(\"Metrics:\", m)\n",
    "\n",
    "# ---------- Aggregate ----------\n",
    "avg_bleu   = np.mean([m[\"bleu\"] for m in metrics_list])\n",
    "avg_meteor = np.mean([m[\"meteor\"] for m in metrics_list])\n",
    "avg_rouge  = np.mean([m[\"rougeL\"] for m in metrics_list])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a6cd1fff-4276-4eea-8822-394c1950516f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 4106/4106 [48:34<00:00,  1.41it/s] \n"
     ]
    }
   ],
   "source": [
    "import random, numpy as np\n",
    "from sacrebleu import BLEU\n",
    "from rouge_score import rouge_scorer\n",
    "from nltk.translate.meteor_score import meteor_score\n",
    "import matplotlib.pyplot as plt\n",
    "from tqdm import tqdm\n",
    "\n",
    "# BLEU scorers for BLEU-1..4\n",
    "bleu1 = BLEU(max_ngram_order=1, effective_order=True)\n",
    "bleu2 = BLEU(max_ngram_order=2, effective_order=True)\n",
    "bleu3 = BLEU(max_ngram_order=3, effective_order=True)\n",
    "bleu4 = BLEU(max_ngram_order=4, effective_order=True)\n",
    "\n",
    "# ROUGE scorer\n",
    "rouge = rouge_scorer.RougeScorer(['rouge1', 'rougeL'], use_stemmer=True)\n",
    "\n",
    "def compute_metrics(ref, pred):\n",
    "    \"\"\"Compute BLEU-n, METEOR, ROUGE-1, ROUGE-L.\"\"\"\n",
    "\n",
    "    # BLEU scores\n",
    "    b1 = bleu1.sentence_score(pred, [ref]).score / 100\n",
    "    b2 = bleu2.sentence_score(pred, [ref]).score / 100\n",
    "    b3 = bleu3.sentence_score(pred, [ref]).score / 100\n",
    "    b4 = bleu4.sentence_score(pred, [ref]).score / 100\n",
    "\n",
    "    # METEOR\n",
    "    ref_tok = ref.split()\n",
    "    pred_tok = pred.split()\n",
    "    meteor = meteor_score([ref_tok], pred_tok)\n",
    "\n",
    "    # ROUGE\n",
    "    r = rouge.score(ref, pred)\n",
    "    rouge1 = r[\"rouge1\"].fmeasure\n",
    "    rougeL = r[\"rougeL\"].fmeasure\n",
    "\n",
    "    return {\n",
    "        \"bleu1\": b1,\n",
    "        \"bleu2\": b2,\n",
    "        \"bleu3\": b3,\n",
    "        \"bleu4\": b4,\n",
    "        \"meteor\": meteor,\n",
    "        \"rouge1\": rouge1,\n",
    "        \"rougeL\": rougeL,\n",
    "    }\n",
    "\n",
    "\n",
    "# Storage lists\n",
    "bleu1_scores, bleu2_scores, bleu3_scores, bleu4_scores = [], [], [], []\n",
    "meteor_scores = []\n",
    "rouge1_scores, rougeL_scores = [], []\n",
    "\n",
    "\n",
    "# Loop through dataset\n",
    "for ex in tqdm(test_data):\n",
    "    dna = ex[\"dna\"]\n",
    "    ref = ex[\"target\"]\n",
    "\n",
    "    pred = model.generate(\n",
    "        dna=dna,\n",
    "        max_new_tokens=80,\n",
    "        device=device,\n",
    "        temperature=0.7,\n",
    "        top_k=50,\n",
    "    )\n",
    "\n",
    "    m = compute_metrics(ref, pred)\n",
    "\n",
    "    bleu1_scores.append(m[\"bleu1\"])\n",
    "    bleu2_scores.append(m[\"bleu2\"])\n",
    "    bleu3_scores.append(m[\"bleu3\"])\n",
    "    bleu4_scores.append(m[\"bleu4\"])\n",
    "\n",
    "    meteor_scores.append(m[\"meteor\"])\n",
    "\n",
    "    rouge1_scores.append(m[\"rouge1\"])\n",
    "    rougeL_scores.append(m[\"rougeL\"])\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bcd3089-60de-4ace-9c4d-35c2e3208d39",
   "metadata": {},
   "outputs": [],
   "source": [
    "avg_bleu = np.mean(bleu_scores)\n",
    "avg_chrf = np.mean(chrf_scores)\n",
    "avg_rouge = np.mean(rouge_scores)\n",
    "\n",
    "print(\"\\n===== AVERAGE METRICS =====\")\n",
    "print(f\"BLEU:    {avg_bleu:.4f}\")\n",
    "print(f\"chrF:    {avg_chrf:.4f}\")\n",
    "print(f\"ROUGE-L: {avg_rouge:.4f}\")\n",
    "\n",
    "\n",
    "plt.figure(figsize=(14,4))\n",
    "\n",
    "plt.subplot(1,3,1)\n",
    "plt.hist(bleu_scores, bins=30, color='skyblue')\n",
    "plt.title(\"BLEU Score Distribution\")\n",
    "\n",
    "plt.subplot(1,3,2)\n",
    "plt.hist(chrf_scores, bins=30, color='lightgreen')\n",
    "plt.title(\"chrF Score Distribution\")\n",
    "\n",
    "plt.subplot(1,3,3)\n",
    "plt.hist(rouge_scores, bins=30, color='salmon')\n",
    "plt.title(\"ROUGE-L Score Distribution\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f72f39d-fb2b-47b2-a51e-19effebc3434",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
